src/init/ClusterManager.ts

```diff
@@ -0,0 +1,120 @@
+import type { Worker } from 'cluster';
+import cluster from 'cluster';
+import { cpus } from 'os';
+import { getLoggerFor } from '../logging/LogUtil';
+import { InternalServerError } from '../util/errors/InternalServerError';
+
+/**
+ * Different cluster modes.
+ */
+enum ClusterMode {
+  /** Scales in relation to `core_count`. */
+  autoScale,
+  /** Single threaded mode, no clustering */
+  singleThreaded,
+  /** Fixed amount of workers being forked. (limited to core_count) */
+  fixed
+}
+
+/**
+ * Convert workers amount to {@link ClusterMode}
+ * @param workers - Amount of workers (typically passed through cli parameter)
+ * @returns ClusterMode enum value
+ */
+function toClusterMode(workers: number | string): ClusterMode {
+  if (workers < -1) {
+    throw new InternalServerError('Invalid workers value (should be [-1, num_cpu]).');
+  }
+  if (workers === -1 || workers === 0) {
+    return ClusterMode.autoScale;
+  }
+  if (workers === 1) {
+    return ClusterMode.singleThreaded;
+  }
+  return ClusterMode.fixed;
+}
+
+/**
+ * This class is responsible for deciding how many affective workers are needed.
+ * It also contains the logic for respawning workers when they are killed by the os.
+ *
+ * The workers values are interpreted as follows:
+ * value | actual workers |
+ * ------|--------------|
+ * `-1` | `num_cores - 1` workers _(autoscale)_ |
+ * `0` | `num_cores` workers _(autoscale)_ |
+ * `1` | `single threaded mode` _(default)_ |
+ * `2` | `2` workers (but never more than `num_cores`) |
+ * `n` | `n` workers (but never more than `num_cores`) |
+ */
+export class ClusterManager {
+  private readonly logger = getLoggerFor(this);
+  private readonly workers: number;
+  private readonly clusterMode: ClusterMode;
+  private counter = 0;
```

Comment:
This can just be a variable in `spawnWorkers` since it's not used anywhere else.